azurerm/resource_arm_application_gateway.go

```diff
@@ -0,0 +1,1697 @@
+							Type:      schema.TypeString,
+							Required:  true,
+							Sensitive: true,
+						},
+
+						"public_cert_data": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceArmApplicationGatewayCreate(d *schema.ResourceData, meta interface{}) error {
+	armClient := meta.(*ArmClient)
+	client := armClient.applicationGatewayClient
+
+	log.Printf("[INFO] preparing arguments for AzureRM ApplicationGateway creation.")
+
+	name := d.Get("name").(string)
+	location := d.Get("location").(string)
+	resGroup := d.Get("resource_group_name").(string)
+	tags := d.Get("tags").(map[string]interface{})
+
+	// Gateway ID is needed to link sub-resources together in expand functions
+	gatewayID := fmt.Sprintf(
+		"/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/applicationGateways/%s",
+		armClient.subscriptionId, resGroup, name)
+
+	properties := network.ApplicationGatewayPropertiesFormat{}
+	properties.Sku = expandApplicationGatewaySku(d)
+	properties.SslPolicy = expandApplicationGatewaySslPolicy(d)
+	properties.GatewayIPConfigurations = expandApplicationGatewayIPConfigurations(d)
+	properties.FrontendPorts = expandApplicationGatewayFrontendPorts(d)
+	properties.FrontendIPConfigurations = expandApplicationGatewayFrontendIPConfigurations(d)
+	properties.BackendAddressPools = expandApplicationGatewayBackendAddressPools(d)
+	properties.BackendHTTPSettingsCollection = expandApplicationGatewayBackendHTTPSettings(d, gatewayID)
+	properties.HTTPListeners = expandApplicationGatewayHTTPListeners(d, gatewayID)
+	properties.Probes = expandApplicationGatewayProbes(d)
+	properties.RequestRoutingRules = expandApplicationGatewayRequestRoutingRules(d, gatewayID)
+	properties.URLPathMaps = expandApplicationGatewayURLPathMaps(d, gatewayID)
+	properties.AuthenticationCertificates = expandApplicationGatewayAuthenticationCertificates(d)
+	properties.SslCertificates = expandApplicationGatewaySslCertificates(d)
+
+	if _, ok := d.GetOk("waf_configuration"); ok {
+		properties.WebApplicationFirewallConfiguration = expandApplicationGatewayWafConfig(d)
+	}
+
+	gateway := network.ApplicationGateway{
+		Name:     azure.String(name),
+		Location: azure.String(location),
+		Tags:     expandTags(tags),
+		ApplicationGatewayPropertiesFormat: &properties,
+	}
+
+	_, errChan := client.CreateOrUpdate(resGroup, name, gateway, make(chan struct{}))
+	err := <-errChan
+	if err != nil {
+		return errwrap.Wrapf("Error Creating/Updating ApplicationGateway {{err}}", err)
+	}
+
+	read, err := client.Get(resGroup, name)
+	if err != nil {
+		return errwrap.Wrapf("Error Getting ApplicationGateway {{err}}", err)
+	}
+	if read.ID == nil {
+		return fmt.Errorf("Cannot read ApplicationGateway %s (resource group %s) ID", name, resGroup)
+	}
+
+	d.SetId(*read.ID)
+
+	log.Printf("[DEBUG] Waiting for ApplicationGateway (%s) to become available", name)
+	stateConf := &resource.StateChangeConf{
```

Comment:
It looks like the Terraform helper + Go SDK does this - so removed the extra code.